# Appendix C — Introduction to Python

It's time for you to learn the basics of Python.
Don't freak out, this is not a big deal.
You're not becoming a programmer or anything,
you're just learning how to use a really good calculator.
Just like knowing how to use a calculator is helpful for doing lots of arithmetic operations,
learning Python is helpful when you need to deal with repetitive operations and procedures.

The Python commands are similar to the commands you give to a calculator.
Using Python requires learning about Python variables, expressions, functions, etc. which are different types of commands that Python understands.
The same way a calculator has different buttons for the various arithmetic operations,
the Python language has a number of commands you can "run" or "execute."
Whereas calculators allow only simple arithmetic calculations of one expression (evaluated when you press = button),
the Python prompt accepts entire "paragraphs" of commands so you can write down more complicated multi-step calculations.
This is what people call "coding".

Coding a.k.a. programming, software engineering, or software development is a much broader topic that is out of scope for this short tutorial. See the article [What is code?](https://www.bloomberg.com/graphics/2015-paul-ford-what-is-code/) if you're interested. Think mobile apps, web apps, APIs, algorithms, CPUs, GPUs, TPUs, SysOps, etc. There is a lot to learn about applications enabled by learning basic coding skills, it's almost like reading and writing skills.

In this tutorial,
we focus on using Python for data calculations and procedures needed for statistics.
Once you learn a bit of Python (this tutorial),
you'll have access to the best-in-class Python libraries
for data management (Pandas, see [intro_to_pandas.ipynb](./intro_to_pandas.ipynb))
and data visualization (Seaborn, see [data_visualization_with_seaborn.ipynb](./data_visualization_with_seaborn.ipynb)).
Learning a few basic Python constructs like the `for` loop
will enable you to simulate frequentist probability calculations
and experimentally verify how statistics procedures work.
This is a really big deal.
If's good to know the statistical formula and recipes,
but it's even better when you can run your own simulations and check when the formulas work and when they fail.

Don't worry there won't be any advanced math—just sums, products, exponents, logs, and square roots. Nothing fancy, I promise. If you've ever created a formula in a spreadsheet, then you're familiar with all the operations we'll see. In a spreadsheet formula you'd use `SUM(` in Python we wrire `sum(`. You see, it's nothing fancy. Yes, there will be a lot of code (paragraphs of Python commands) in this tutorial, but you can totally handle this. If you ever start to freak out an think "OMG this is too complicated!" remember that Python is just a fancy calculator.

**Code cells run Python commands**
The Python command prompt (each of the code cells in this notebook)
allows you to enter Python commands and "run" them by clicking the play button
in the toolbar or pressing SHIFT + ENTER.
You can make Python calculate the value of 10 raised to the power 5 by entering `10**5` in a code cell.


## Table of contents
- Basics
- Variables and expressions
- Lists and for loops
- Functions
- Dictionaries and other data structures
- Classes and objects
- Syntax review

- Python libraries:
  - Best of stdlib: math, random, statistics, json, csv
  - Numerical computing: numpy
  - Scientific computing: scipy
  - Extracting web content: requests and BeautifulSoup
  - Pandas (see [intro_to_pandas.ipynb](./intro_to_pandas.ipynb))
  - Seaborn (see [data_visualization_with_seaborn.ipynb](./data_visualization_with_seaborn.ipynb))  

### Bonus topics (if time)

  - Command line: similar to how in computing user interfaces click (or  double click) means "run <program>", a command line interface allows us to run programs by simply typing out their name, as in    program    and pressing. Enter. This is where the notion of "calling a program/function" comes from, you   just write it's name in a command line prompt and this is equivalent to "calling it" (making it run, just click clicking on it)
  - Command line scripts: run any python code on command line using  python
  myscript.py (or if configured as an executable script, simply ./myscript.py)
  - Documentation: READMEs, __doc__ strings, and other technical altruism basics
  - Debugging (intro to JupyterLab debugger)
  - Code Testing (e.g. test_fun: a function that checks that `fun` returns the expected outputs on some set of inputs)
  - Algorithms: different approaches for solving computational problem

# Basics

Programming consists of writing down instructions for a computer. The instructions are also called "commands", "code", "source code", etc. Since this tutorial is about Python, all the instructions in the boxes you'll see below are written in the Python programming language. 

You can "run" the code in a given box by pressing the play button on the left side, or by pressing _Shift_ and _Enter_ keys at the same time. When you run a code cell, you're telling the computer to "execute" the Python instructions in that cell, which means to do the actions described in the code, and print the final value computed in that cell.

Running a code cell is similar to using the EQUALS button on the calculator: whatever math expression you entered, the calculator will compute its value and display it as the output. The process is identical when you execute some Python code, but you're allowed to input multiple lines of commands at once. The computer will execute the lines of code one by one in the order it sees them.

The result of final calculations in the cell gets automatically printed in the output cell right below the input cell. This feature allows you to skip the print statements, since the last output gets printed automatically for you. This makes it easy and fun to explore and "poke around" each example given in this notebook. Click the `[+ Code]` button below any cell to add a blank cell where you can try inputting some expressions and running them by pressing _Shift_ and _Enter_.



### Errors
Sometimes the executed commands will cause an error, and in these cases Python will give an error message describing the problem encountered. Get psychologically ready for those, because they can be very discouraging. REJECTED! The computer doesn't like what you entered. SyntaxError, ValueError, etc. The error messages look scary, but really they are there to help you—if you read what the error message tells you, you will know what needs to be fixed in your input. The error message literally describes the problem!

In [1]:
3/0

ZeroDivisionError: division by zero

Here we get an error since we're trying to compute an expression that contains a divide by zero error: Python tell us a `ZeroDivisionError: division by zero` has occurred, i.e., check your math expressions!

# Variables and expressions




## Variables

Similar to variables in math, a variable in Python is a convenient name we use to refer to any value: a onstant `c`, the input to a funciton `x`, the output of a function `y`, or any other intermediate value.

To assign a value to a variable, you use the symbol `=` as follows, from left to right:

* we start by writing the name of the variable
* then, we add the symbol `=`
* finally, we write the value of the variable

In [2]:
x = 3
print(x)

3


In the first line of the above code cell, we set the variable `x` to the value `3`. On the second we call the `print` function to display the value of x to the screen.

### Variables types

There are multiple types of variables in Python:

- **int** - integers ex: `34`,`65`, `78`, `-4`, etc. (rougly equivalent to $\mathbb{Z}$)
- **float** - ex: `4.6`,`78.5`, `1e-3` (full name is "floating point number"; similar to $\mathbb{R}$ but only with finite precision)
- **bool** a Boolean truth value with only two choices: `True` or `False`.
- **string** - text ex: `'Hello'`, `'Hello everyone'`
- **list** a sequence of values ​​- ex: `[69, 81, 92, 77]`. The beginning and the end of the list are denoted by the brackets `[` and `]`, and its elements are separated by commas.
- **dictionary** a collection of key-value pairs. Each key is associated with a value - ex: `{'first_name': 'Julie', 'last_name': 'Tremblay', 'score': 98}`. Dictionaries are denoted by curly braces `{` and `}` inside which we place `'key': value`, pairs separated by commas.
- **tuples**, **sets**, **functions**, etc. These are other useful Python building blocks which we'll talk about in later sections.

Let's look at some examples with variables of different types:
an `int`eger, a `float`ing point number, a `bool`ean value, a `str`ing,
a list, and a `dict`ionary.

In [3]:
score = 98
average = 77.5
above_the_average = True
message = "Hello everyone"
scores = [61, 85, 92, 72]
profile = {"first_name":"Julie", "last_name":"Tremblay", "score":98}

In [4]:
len(message)

14

In [5]:
message.split()

['Hello', 'everyone']

You can explore the different methods available on any python object `int`, `float`, `str`, etc.  by starting to type the dot `.` after the name, e.g., `message.` then pressing the TAB button to get an "autocomplete" dropdown of all the methods available on the variable `message`. Most of these methods are common to all strings in Python.

## Expressions

Similar to expresisons in algebra, a Python expression is can be any combination of variables and operations:

In [6]:
# Expression involving numerical values
secs_in_1min = 60 
secs_in_1day = secs_in_1min * 60 * 24
secs_in_1week = secs_in_1day * 7
print("The number of seconds in a week is", secs_in_1week)


# Expression involving strings
name = "Julie"
message = "Hello " + name    # for strings, + means concatenate
print(message)


# Expression involving a list
scores = [61, 85, 92, 72]
average = sum(scores)/len(scores)
#        `sum` computes the sum of values in the list
#                and `len` gives you the length of the list
print("The average score is", average)


# String expression using a values from a dictionary
profile = {"first_name":"Julie", "last_name":"Tremblay", "score":98}
message2 = "Hi " + profile["first_name"]
print(message2)

The number of seconds in a week is 604800
Hello Julie
The average score is 77.5
Hi Julie


Note in all the above examples, the code had the form:

```Python
var_name = <some expression>
```
which is the important new pattern you have to get used to in programming. Even though this looks like a math equation, the meaning you have to associate with it is much simpler—we are setting variable `var_name` to the value of the expression `<some expression>`.

Don't worry about the lists and dictionary examples—I know they are complicated and we haven't explained all the syntax. We'll get to that in just a little bit. First let's practice computing some Python expresions.

In [7]:
# Numeric expressions

expr1 = 1 + 2.4
print(expr1)  # 3.4

expr2 = 4 - 6
print(expr2)  # -2

expr3 = 0.5 * 3
print(expr3)  # 1.5

3.4
-2
1.5


Here is an expression that Python cannot compute,
so it raises an exception.

In [8]:
expr4 = 5/0

ZeroDivisionError: division by zero

You'll see these threatening looking messages on a red background any time Python encounters an error when trying to run the commands you specified.
This is nothing to be alarmed by.
It usually means you made a typo (symbol not defined error),
forgot a syntax element (e.g. `(`, `,`, `[`, `:`, etc.),
or tried to compute something impossible like the `ZeroDivisionError` in the above example.

The way to read these red messages is to focus on the name of the exception and the message that gets printed on the last line. This should tell you what you need to fix. The solution will be obvious for typos and syntax errors, but for more complicated situations requires googling and searching on stack overflow.


### Integration exercise

Expression for converting a temperature in Celsius to temperature in Fahrenheit.


In [9]:
temp_C = 100
temp_F = (temp_C * 9/5) + 32

temp_F


212.0

Test: if temp_C = 100 then value of expression should be temp_F = 212.

### String expressions

In [10]:
message3 = "Hello" + " " + "world"  # + means concatenate
print(message3)

message4 = "Ai " * 3                # * means repeat
print(message4 + 'Caramba!') 

Hello world
Ai Ai Ai Caramba!


### Strings are lists of characters

In [11]:
# Define a str of length 26 that contains all the lowercase Latin letters
letters = "abcdefghijklmnopqrstuvwxyz"

# Accessing individual characters within a string
print("The index of the letter `a` in the string letters is 0")
first_letter = letters[0]  # a
print(first_letter)

print("The index of the letter `b` in the string letters is 1")
second_letter = letters[1]  # b
print(second_letter)

print("The index of the last letter in the string letters is -1")
last_letter = letters[-1]  # z
print(last_letter)

print("The last element in list of 26 elments has index 25")
print(letters[25] == "z")

print('\n\n\n')  # '\n' is a special character (an escape sequence) that prints a newline
                 # we'll use this kind of preint-nelines statements to logically
                 # separate the out outplut lines

# Slicing = getting the substring for a particular range of indices
first_four = letters[0:4]
print('The first four letters of the alphabet are:', first_four)
# the notation 0:4 is sugar syntax for `slice(0,4)` and corresponds
# to the range of indices for 0 to 4 (non-inclusinve): [0,1,2,3].

The index of the letter `a` in the string letters is 0
a
The index of the letter `b` in the string letters is 1
b
The index of the last letter in the string letters is -1
z
The last element in list of 26 elments has index 25
True




The first four letters of the alphabet are: abcd


### Boolean expressions

You can use `bool` variables and the logical operations `and`, `or`, `not`, etc. to build more complicated boolean expressions (disjunctions, conjunctions, negations, etc.).

In [12]:
print('not True ==', not True)
print('not False ==', not False)
print('True and True ==', True and True)
print('True and False ==', True and False)
print('True or False ==', True or False)
print('False or False ==', False or False)

not True == False
not False == True
True and True == True
True and False == False
True or False == True
False or False == False


## Types and type conversions

The function `type` tells you the type of any variable, meaning what kind of number or object it is.
Look back to the list above—the Python types are shown in **bold**.

In [13]:
# integrers
score = 98
type(score)

int

In [14]:
average = 77.5
type(average)

float

In [15]:
above_the_average = True
type(above_the_average)

bool

You can convert between each of these types using the function which has the same name as the type of an object

- `int` : transform any expression into an int
- `float`: transform any expression into a flaot
- `bool`: transform any expression into a `True` or `False`
- `str`: transform an expression to it's string representaiton (i.e. this is what the function `print` does).

### Useful anything-to-int conversions

In [16]:

# str --> int
number_as_str = '65'
print(number_as_str, 'has type', type(number_as_str))
number = int(number_as_str)
print(number, 'has type', type(number))

#  float --> int
precise_number_as_float = 3.2
print(precise_number_as_float, 'has type', type(precise_number_as_float))
rounded_number = int(precise_number_as_float)
print(rounded_number, 'has type', type(rounded_number))

#  bool --> int
true_value = True
false_value = False
print('When converted to an int, True is', int(True), 'and False is', int(False))

65 has type <class 'str'>
65 has type <class 'int'>
3.2 has type <class 'float'>
3 has type <class 'int'>
When converted to an int, True is 1 and False is 0


### Useful anything-to-float conversions

In [17]:
# str --> float
print(float('1.2'))  # note using decimal dot . not decimal , as in Europe
print(float('1e6'))  # one million = 1000000 = 1x10^6. The e is shorthand for x10^
print(float('4'))

# int/int --> float autoconversion
print('If you divide an integrer by an integer in Python using the / operator...')
print('...you get a float number', 6/5, 'has type', type(6/5))

print('There is also an divistion without autoconversion operator // which ...')
print('...you get an integer', 6//5, 'has type', type(6//5))

1.2
1000000.0
4.0
If you divide an integrer by an integer in Python using the / operator...
...you get a float number 1.2 has type <class 'float'>
There is also an divistion without autoconversion operator // which ...
...you get an integer 1 has type <class 'int'>


### Common anything-to-str conversions

In [18]:
# use the str function
str1 = str(42)
str2 = str(4.2)
str3 = str(True)
str4 = 'a string'
print('string representations', str1, str2, str3, str4)

# float -> str
import math
pi = math.pi  # constant equal to ratio of circumference to diameter of a circle
print('str(math.pi) =', str(math.pi))  # defaults to max precision

# control precision and print formatting using format. For more info, see
# https://python-reference.readthedocs.io/en/latest/docs/functions/format.html
print('pi to two decimals', '{:.2f}'.format(pi))
print('pi to seven decimals', '{:.7f}'.format(pi))

string representations 42 4.2 True a string
str(math.pi) = 3.141592653589793
pi to two decimals 3.14
pi to seven decimals 3.1415927


### Converting expressions to boolean values

Up until now we were in the land of math and text manipulation of numbers, which all make sense intuitively. 

Next we'll talk about the boolean values of Python expressions. This will feel a little weird at first since we're forcing some arbitray Python object (could be a number, a string, a list, a dictionary, etc) and asking the questions is the value `True` or `False` ?

There is a specific convention about which values are considered "truthy" (i.e. get converted to `True` when converted to boolean using `bool`) and which expressions are falsy (i.e. get converted to `False` when passed through `bool`). 

I know this seems like boring details, but please read the next example carefully because "truthyness" and "falsyness" will play a big role in coding: every time you use an if or elif statement in Python, we are implicitly calling `bool` on an expression so it's a good idea to get to know what `bool` does to different types of variables.

In [19]:
# int --> bool
print('Any non-zero integer is considered as True')
print(bool(1), bool(-2), bool(10000))
print('Zero is False')
print(bool(0), bool(-0), bool(int('0')))

print('\n')

# float --> bool
print('Any non-zero float is considered as True')
print(bool(1.0), bool(-2.0), bool(10000.0))
print('Zero is False')
print(bool(0.0), bool(-0.0), bool(float('0.0')))


# str --> bool
print('Any non-empty string is considered True')
print(bool('as'), bool(''))

print('\n')

# list --> bool
print('Any non-empty list is considered True')
print(bool([1]), bool([1,2]), bool(range(0,10000)))
print('Empty list is considered False')
print(bool([]), bool(list()), bool(list('[]')))

print('\n')

# dict --> bool
print('Any non-empty dict is considered True')
print(bool({1:11}), bool({1:11,-2:22}))
print('Empty dict is considered False')
print(bool({}), bool(dict()))

Any non-zero integer is considered as True
True True True
Zero is False
False False False


Any non-zero float is considered as True
True True True
Zero is False
False False False
Any non-empty string is considered True
True False


Any non-empty list is considered True
True True True
Empty list is considered False
False False True


Any non-empty dict is considered True
True True
Empty dict is considered False
False False


In [20]:
bool(2)

True

## Your turn to try this...

Try typing in some Python code in this cell. If you've been simply reading until now, this is your chance to switch to "active" mode: use the rocket-button in the top right of the menu at the top, and choose the `Live Code` option to make all the cells in this notebook interactive, then try entering some Python commands in this code cell below:




## Exercises

- Review Python syntax cheatsheet [https://blog.finxter.com/wp-content/uploads/2020/07/Finxter_WorldsMostDensePythonCheatSheet.pdf](https://blog.finxter.com/wp-content/uploads/2020/07/Finxter_WorldsMostDensePythonCheatSheet.pdf)
  - add single dot next to concepts you've hear about
  - double dot next to python concepts you understand
  - triple dot next to concepts you've used in your code
- Try poking-around and explore expressions involving numbers (`int` and `float`),
  strings (`str`), and booleans (`bool`).
- Go through all quiz questions in reading material as notebooks:
  - 03-Variables: [https://introductorypython.github.io/tutorials/03-Variables.html](https://introductorypython.github.io/tutorials/03-Variables.html)
  - 04-Operarators: [https://introductorypython.github.io/tutorials/04-Operators.html](https://introductorypython.github.io/tutorials/04-Operators.html)
  - 06-Data types: [https://introductorypython.github.io/tutorials/06-DataTypes.html](https://introductorypython.github.io/tutorials/06-DataTypes.html) 
  - Review Collections: Lists section. This is important because we can use lists
    to represent vectors in Python, for example the two-dimensional vector can be defined as `v = [3,2]`
  - Complete the section on for loops in 07-Loops notebook.
    The for loop is important because it allows you to do operations for each element in the list.

# Lists and for loops


To create a list of two values:
- start with an opening square bracket `[` ,
- then put the first value,
- comma `,`,
- then the second value,
- finally close the square bracket `]`

In [82]:
first_val = 3
second_val = 4

my_list = [first_val, second_val]
my_list

[3, 4]

In [21]:
scores = [61.0, 85.0, 92.0, 72.0]  # define a list of floats
scores

[61.0, 85.0, 92.0, 72.0]

In [22]:
# lists have a "length"
len(scores)

4

In [23]:
# elements of a list are acccessed using [ ] and 0-based index
scores[0]  # first score

61.0

In [24]:
# lists can be sorted
sorted(scores)  # returns a new list of sorted scores

[61.0, 72.0, 85.0, 92.0]

In [25]:
scores

[61.0, 85.0, 92.0, 72.0]

In [26]:
scores.sort()  # in-place sort the list
scores

[61.0, 72.0, 85.0, 92.0]

In [27]:
scores.reverse()
scores

[92.0, 85.0, 72.0, 61.0]

In [28]:
scores.append(22)

In [29]:
scores

[92.0, 85.0, 72.0, 61.0, 22]

In [30]:
scores.insert(2, 25)
scores

[92.0, 85.0, 25, 72.0, 61.0, 22]

In [31]:
scores.pop()

22

Just like `.sort()` method, lists have all kinds of useful methods `.insert`, `.remove`, `.pop`, `.reverse`, ...  

You can see all those methods by starting to type `scores.` then pause for a second to see the auto-complete suggestions:

In [32]:
# scores.

In [33]:
scores = [61.0, 85.0, 92.0, 72.0]
scores.sort()

## For loops

The "for loop" is a Python code construct of the form:
```Python
for el in <container>:
    <operations on element `el`>

```
that allows to repeat a block of operations **for each element of the list**.

In [34]:
# Example 1: print all the scores
for score in scores:
    print(score)

61.0
72.0
85.0
92.0


In [100]:
# Example 2: compute the average score  ==  sum(scores)/len(scores)

total = 0

for score in scores:
    total = total + score

total / len(scores)

77.5

The name of the variable used for the for loop is totally up to you, but in general you should choose logical names for elements of the list.

Here is a for loop that uses the single-letter variable:

In [36]:
for s in scores:
    print(s)


61.0
72.0
85.0
92.0


In [37]:
scores

[61.0, 72.0, 85.0, 92.0]

In [38]:
enumerate(scores)

In [111]:
# Bonus concept: use `enumerate` to get pairs (index, value) from a list
# enumerate(scores) == [(0, 61.0), (1, 72.0), (2, 85.0), (3, 92.0)]

# example
for idx, score in enumerate(scores):
    # this for loop has two variables index and score
    print("Processing score", score, "which is at index", idx, "in the list")

Processing score 61.0 which is at index 0 in the list
Processing score 72.0 which is at index 1 in the list
Processing score 85.0 which is at index 2 in the list
Processing score 92.0 which is at index 3 in the list


In [115]:
# New concept: use `zip(list1,list2)` to get pairs (value1, value2) from two lists 
# list(zip([1,2,3], ['a','b','c'])) == [(1, 'a'), (2, 'b'), (3, 'c')]

# example
list1 = [1, 2, 3]
list2 = [4, 5, 6]

for value1, value2 in zip(list1, list2):
    print("Processing values", value1, "and", value2)


Processing values 1 and 4
Processing values 2 and 5
Processing values 3 and 6


In [116]:
list1 = [1, 2, 3]
list2 = [4, 5, 6]

list(zip(list1, list2))

[(1, 4), (2, 5), (3, 6)]

### List functions

Your turn to play with lists now! Complete the code required to implement the functions `compute_mean` and `compute_std` below.


#### Question 1: Mean

The formula for the mean of a list of numbers $[x_1, x_2, \ldots, x_n]$ is:
$$
    \text{mean} = \overline{x}
    = \frac{1}{n}\sum_{i=1}^n x_i
    = \tfrac{1}{n} \left[ x_1 + x_2 + \cdots + x_n \right].
$$


Write the function `mean(numbers)`: a function that computes the mean of a list of numbers

In [96]:
def mean(numbers):
    """
    Computes the mean of the `numbers` list using a for loop.
    """
    total = 0
    for number in numbers:
        total = total + number
    return total / len(numbers)  


mean([100,101])

100.5

In [43]:
# TEST CODE (run this code to test you solution)

def random_list(n=10, min=0.0, max=100.0):
    """Returns a list of length `n` of random floats between `min` and `max`."""
    import random
    values = []
    for i in range(n):
        r = random.random()
        value = min + r*(max-min)
        values.append(value)
    return values


def test_compute_mean(function):
    """
    Run a few lists to check if value returned by `function` matches expected.
    """
    import math, statistics
    assert function([1,1,1]) == 1
    assert function([61,72,85,92]) == 77.5
    list10 = random_list(n=10)
    assert math.isclose(function(list10), statistics.mean(list10))
    list100 = random_list(n=100)
    assert math.isclose(function(list100), statistics.mean(list100))
    print("All tests passed. Good job y'all!")


# RUN TESTS
test_compute_mean(compute_mean)

All tests passed. Good job y'all!


In [44]:
(1 + 1e-15)  ==  1 

False

In [45]:
import math
math.isclose(1 + 1e-10, 1)

True

#### Question 2: Sample standard deviation

The formula for the sample standard seviation of a list of numbers is:
$$
    \text{std}(\textbf{x}) = s
    = \sqrt{ \tfrac{1}{n-1}\sum_{i=1}^n (x_i-\overline{x})^2 }
    = \sqrt{ \tfrac{1}{n-1}\left[ (x_1-\overline{x})^2 + (x_2-\overline{x})^2 + \cdots + (x_n-\overline{x})^2\right]}.
$$

Note the division is by $(n-1)$ and not $n$. Strange, no? You'll have to wait until stats to see why this is the case.

Write `compute_std(numbers)`: computes the sample standard deviation

In [93]:
import math

def std(numbers):
    """
    Computes the sample standard deviation (square root of the sample variance)
    using a for loop.
    """
    mean = compute_mean(numbers) 
    total = 0
    for number in numbers:
        total = total + (number-mean)**2
    var = total/(len(numbers)-1)    
    return math.sqrt(var)

numbers = list(range(0,100))
std(numbers)

29.011491975882016

In [94]:
# compare to known good function...
import statistics
statistics.stdev(numbers)

29.011491975882016

In [95]:
# TEST CODE (run this code to test you solution)

def test_std(function):
    """
    Run a few lists to check if value returned by `function` matches expected.
    """
    import math, statistics
    assert function([1,1,1]) == 0
    assert math.isclose(function([61,72,85,92]), 13.771952173409064)
    list10 = random_list(n=10)
    assert math.isclose(function(list10), statistics.stdev(list10))
    list100 = random_list(n=100)
    assert math.isclose(function(list100), statistics.stdev(list100))
    print("All tests passed. Good job y'all!")


# RUN TESTS
test_std(std)

All tests passed. Good job y'all!


### If statements

In [62]:
if True:
    print("This code will run")
    print("This code will run too")
    
if False:
    print("This code will not run")

This code will run
This code will run too


Can construct boolean values like `True` and `False` using various expressions that involve comparison,
`==` equal to, `>` greater than, `<` less than, etc.

In [63]:
x = 3

x > 2

True

In [64]:
if x > 2:
    print("x is greater than 2")
else:
    print("x is less than or equal to 2")

x is greater than 2


We can do multiple checks using `elif` statements.

In [65]:
temp = 20

if temp >= 100:
    print("Gas")
elif 0 < temp and temp < 100:
    print("Liquid")
else:
    print("Solid")

Liquid


### Things to discuss

In [66]:
## 1. imports
from math import acos
acos(1)
# vs
import math
math.acos(1)



## 2. list membership
v = [3,2]
len(v) == range(2,10)



## 3. joint-iteration over two lists using zip


## 4. float(cross) error


## 5. syntax for defining vs. calling functions



False

# Functions

Functions are one of the most important concepts in math and programming.


    y = f(x)  # common convention in math to call function inputs x, and outputs y

We can also draw as `x -----f----> y` the function is a map from input values x to an output value y


    def f(x):
        <steps to compute y from x>
        return y


Functions!! Finally we get to the good stuff! The previous two sessions were important
foundations, but now we get to unlocking the first superpower — modelling.
Once you know the basic properties of 10 or so functions,
you can build precise mathematical models for any real-world system.


Functions are all over the place:
- In high school math (the green book) we learn the basic vocabulary of y=f(x) functions
  and their parameters, which allows us to describe any real world process
- In calculus we analyze functions f(x) behaviour over time
  (integral of f = sum of values of f(x) between x=start and x=finish;
  and derivative of f at a = the slope of the graph of f(x) when x=a)
- In linear algebra we study linear transformations, which are functions
  that satisfy f(ax+by)=af(x)+bf(y), meaning a linear combination of inputs
  produces the same linear combination of outputs.
- In probability theory we use functions to describe the probability density of
  a random variables. For example X = Normal(mu, sigma^2) is a random variable
  whose density is described by the function p(x) = K*exp(-((x-mu)/sigma)^2/2)
  in stats we talk about functions computed from samples (estimators)
- In ML we learn about probabilistic models and use them to predict y from given input x

## Python functions

Functions in Python are similar to functions in math: a transformation that takes certain inputs and produces certain outputs. The math functions you're familiar with take numbers as inputs and produce numebrs as outputs, but a Python function can take on any type of input and produce any type of output.

Functions allow us to build chunks of reusable code that we can later reuse in other programs.

We declare a function with the keyword `def`.

```python
def function_name(function inputs):
    """
    doc string that describes what the function does (optional)
    """
    <function body line 1>
    <function body line 2>
    <function body line ...>
    <function body line n-1>
    return <function output value>
```

We enter the name of the function, then define the functions arguments inside parentheses—the names of the variables that the function receives as inputs. Then the function body is written as an indented block of code (all lines start with four spaces indentation). The output of the function is specified using the `return` keyword. The return statement is usually the last line in the function body.

Certain functions do not return a value (we call these _procedures_) and they consist of sequences of commands we want to execute, that don't have any outputs. FunctIons can also be attached to objects, in which case they are called _methods_. We'll talk about these layer on, for now let's focus on simple math-like functions that receive some input and produce output:

#### Example 1

A first example of a simple math-like function. The function is called `f`,
takes numbers as inputs, and produces numbers as outputs:

In [67]:
def f(x):
    return 2*x + 3

f(10)

23

#### Example 2

In [68]:
# Example from Session 1
import random

def flip_coin():
    r = random.random()
    if r < 0.5:
        print("heads")
    else:
        print("tails")

flip_coin()  # no return value, but prints output

tails


#### Example 3

Write a function `water_phase(temp)` that takes input temperature `temp` in Celcius,
uses if/else statements to find what state water is in (assume pressure is 1atm).
The function returns a string, which is one of "Solid", "Liquid", "Gas".

In [69]:
def water_phase(temp):
    """
    Returns phase of water at `temp`.
    Input temp is temperature in Celcius (int or float)
    temp must be greater than −273.15.
    """
    if temp > 0 and temp < 100:
        return 'Liquid'
    elif temp <= 0:
        return 'Solid'
    elif temp >= 100:
        return 'Gas'

In [70]:
# help(water_phase)
water_phase?

Signature: water_phase(temp)
Docstring:
Returns phase of water at `temp`.
Input temp is temperature in Celcius (int or float)
temp must be greater than −273.15.
File:      /var/folders/wc/2r44j8gs4gn56t1xtw5f6wlm0000gn/T/ipykernel_73602/1578336333.py
Type:      function


In [71]:
## tests to try: correct implementation of `water_phase` should return all True
print( water_phase(20.0) == "Liquid" )
print( water_phase(-20.0) == "Solid" )
print( water_phase(200.0) == "Gas" )
print( water_phase(0.0) in ["Liquid", "Solid"] )

True
True
True
True


In [72]:
## range tests
results = []

for temp in range(1, 100):
    result = (water_phase(temp) == 'Liquid')
    results.append(result)

# temp <= 0
for temp in range(-100, 1):
    result = (water_phase(temp) == 'Solid')
    results.append(result)

# temp >= 100
for temp in range(100, 5000):
    result = (water_phase(temp) == 'Gas')
    results.append(result)

print('Completed a total of', len(results), 'checks...')
all(results)  # True if all results are True

Completed a total of 5100 checks...


True

In [73]:
# any = OR for a list
# all = AND for a list
all([True, True, False])

False

#### Exercise 2

Write a Python function called `temp_convert` that converts C to F

In [74]:
import math
def temp_convert(temp_C):
    """
    Convert the temprate temp_C to temp_F.
    """
    pass



### Exercise 4

In [86]:
import random
def roll_die():
    value = random.randint(1, 6)
    return value

In [87]:
roll_die()

3

In [83]:
for n in range(0,10000):
    if roll_die() not in [1,2,3,4,5,6]:
        print("error")